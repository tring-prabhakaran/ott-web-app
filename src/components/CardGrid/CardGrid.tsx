import React, { useEffect, useState } from 'react';
import classNames from 'classnames';
import InfiniteScroll from 'react-infinite-scroller';

import styles from './CardGrid.module.scss';

import useBreakpoint, { Breakpoint, Breakpoints } from '#src/hooks/useBreakpoint';
import { isLocked } from '#src/utils/entitlements';
import Card from '#components/Card/Card';
import type { AccessModel } from '#types/Config';
import type { Playlist, PlaylistItem } from '#types/playlist';
import { parseAspectRatio, parseTilesDelta } from '#src/utils/collection';
import InfiniteScrollLoader from '#components/InfiniteScrollLoader/InfiniteScrollLoader';

const INITIAL_ROW_COUNT = 6;
const LOAD_ROWS_COUNT = 4;

const defaultCols: Breakpoints = {
  [Breakpoint.xs]: 2,
  [Breakpoint.sm]: 2,
  [Breakpoint.md]: 3,
  [Breakpoint.lg]: 4,
  [Breakpoint.xl]: 5,
};

type CardGridProps = {
  playlist: Playlist;
  watchHistory?: { [key: string]: number };
  isLoading: boolean;
  cols?: Breakpoints;
  currentCardItem?: PlaylistItem;
  currentCardLabel?: string;
  accessModel: AccessModel;
  isLoggedIn: boolean;
  hasSubscription: boolean;
  hasLoadMore?: boolean;
  loadMore?: () => void;
  onCardHover?: (item: PlaylistItem) => void;
  getUrl: (item: PlaylistItem) => string;
};

function CardGrid({
  playlist,
  watchHistory,
  isLoading = false,
  cols = defaultCols,
  currentCardItem,
  currentCardLabel,
  accessModel,
  isLoggedIn,
  hasSubscription,
  hasLoadMore,
  getUrl,
  loadMore,
  onCardHover,
}: CardGridProps) {
  const breakpoint: Breakpoint = useBreakpoint();
  const posterAspect = parseAspectRatio(playlist.cardImageAspectRatio || playlist.shelfImageAspectRatio);
  const visibleTiles = cols[breakpoint] + parseTilesDelta(posterAspect);
  const [rowCount, setRowCount] = useState(INITIAL_ROW_COUNT);

  const defaultLoadMore = () => setRowCount((current) => current + LOAD_ROWS_COUNT);
  const defaultHasMore = rowCount * visibleTiles < playlist.playlist.length;

  useEffect(() => {
    // reset row count when the page changes
    setRowCount(INITIAL_ROW_COUNT);
  }, [playlist.feedid]);

  const renderTile = (playlistItem: PlaylistItem) => {
    const { mediaid } = playlistItem;

    return (
      <div className={styles.cell} key={mediaid} role="row">
        <div role="cell">
          <Card
            progress={watchHistory ? watchHistory[mediaid] : undefined}
            url={getUrl(playlistItem)}
            onHover={typeof onCardHover === 'function' ? () => onCardHover(playlistItem) : undefined}
            loading={isLoading}
            isCurrent={currentCardItem && currentCardItem.mediaid === mediaid}
            currentLabel={currentCardLabel}
            isLocked={isLocked(accessModel, isLoggedIn, hasSubscription, playlistItem)}
            posterAspect={posterAspect}
            item={playlistItem}
          />
        </div>
      </div>
    );
  };

  return (
    <InfiniteScroll pageStart={0} loadMore={loadMore || defaultLoadMore} hasMore={hasLoadMore || defaultHasMore} loader={<InfiniteScrollLoader key="loader" />}>
      <div className={classNames(styles.container, styles[`cols-${visibleTiles}`])} role="grid">
        {playlist.playlist.slice(0, rowCount * visibleTiles).map(renderTile)}
      </div>
    </InfiniteScroll>
  );
}

export default CardGrid;
