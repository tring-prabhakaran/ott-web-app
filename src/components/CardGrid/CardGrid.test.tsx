import * as React from 'react';

import CardGrid from './CardGrid';

import playlistFixture from '#test/fixtures/playlist.json';
import { renderWithRouter } from '#test/testUtils';
import type { Playlist, PlaylistItem } from '#types/playlist';

describe('<CardGrid>', () => {
  it('renders and matches snapshot', () => {
    const playlist = playlistFixture as Playlist;
    const { container } = renderWithRouter(
      <CardGrid
        playlist={playlist}
        onCardHover={vi.fn()}
        isLoading={false}
        accessModel={'SVOD'}
        isLoggedIn={true}
        hasSubscription={true}
        getUrl={(item: PlaylistItem) => `m/${item.mediaid}`}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
