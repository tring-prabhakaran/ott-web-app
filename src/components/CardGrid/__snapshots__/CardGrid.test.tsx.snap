// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CardGrid> > renders and matches snapshot 1`] = `
<div>
  <div>
    <div
      class="_container_7175d1 _cols-5_7175d1"
      role="grid"
    >
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label="Agent 327"
            class="_card_d75732"
            href="/m/uB8aRnu6"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    4 min
                  </div>
                </div>
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              >
                Agent 327
              </div>
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label="Big Buck Bunny"
            class="_card_d75732"
            href="/m/awWEFyPu"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    10 min
                  </div>
                </div>
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              >
                Big Buck Bunny
              </div>
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label="Elephants Dream"
            class="_card_d75732"
            href="/m/eFPH2tVG"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    11 min
                  </div>
                </div>
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              >
                Elephants Dream
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
